### DISCUSSION TITLE
Use of libdragon's RDP validator tool

### DISCUSSION THREAD

**Discussion by kev4cards** [gonetz/GLideN64#2795] [Ideas]
@gonetz Have you considered using libdragon's RDP validator tool that is available in its unstable branch (https://github.com/DragonMinded/libdragon/tree/unstable/tools/rdpvalidate)? It has been quite useful in homebrew development as well as exposing graphics emulation bugs. Rasky has put tremendous effort into making it work well.

**Comment by gonetz** [comment]
"This tool disassembles and validates a sequence of RDP commands provided in binary or hex format.\n"
I can see that it can be useful for homebrew development, but how can it be used to expose graphics emulation bugs?

**Reply by kev4cards** [reply]
It should just be a matter of dumping the RDP command sequence and running it through the validator. It is my understanding that the validator runs at the hardware level and operates as a state machine simulator. That said, I really do not know the low-level details of the tool. I think the validator makes use of the code here: https://github.com/DragonMinded/libdragon/blob/unstable/src/rdpq/rdpq_debug.c .

**Reply by gonetz** [reply]
How can it help to expose graphics emulation bugs? The RDP commands come from emu core. The command sequence can be 100% valid, and it is valid in most cases. However, we still have graphics issues, because there are bugs in emulation of these commands on graphics plugin side.  Homebrew developers really can use this tool to validate their work, but I don't see how I, as a graphics plugin developer, can get use of it.

**Reply by kev4cards** [reply]
Yeah. After conversing with Rasky, the validator tool can be useful for catching undefined behaviors. That said, I would not be surprised if you have not handled most of those already.
I know the plugin has many outstanding bugs, and emulation of the RDP has been tough. So, I thought this tool would be helpful, but apparently, it is- most likely- only mildly useful for emulation purposes.

**Reply by weinerschnitzel** [reply]
https://github.com/Themaister/parallel-rdp#project-structure

There are some conformance and replay tools in parallel-rdp, using Angrylion as the reference standard. 

Perhaps GLideN64 can have similar tools that can pinpoint bugs
